### ISSUE TITLE
support for TEST_REQUIRES ?

### ISSUE THREAD

**Issue by fbrosson** [haikuports/haikuporter#151]
When a recipe has a `TEST` function which needs commands that are not needed during `BUILD` we still have to add them to `BUILD_PREREQUIRES`.
It would be nice to be able to take out of `BUILD_PREREQUIRES` all the commands which are not required for `BUILD` and "move" them to a separate variable, say, TEST_REQUIRES or TEST_PREREQUIRES.

Suppose that we need md5sum and sha256sum for running `TEST`, and that we don't need them in `BUILD`. We would add this after `BUILD_PREREQUIRES`:
```
TEST_REQUIRES="
	cmd:md5sum
	cmd:sha256sum
	"
```
Then, we would (currenly need to) add this hack:
```
if [ "$recipeAction" = TEST ]; then
	BUILD_PREREQUIRES="$BUILD_PREREQUIRES
		$TEST_REQUIRES
		"
fi
```
It would be cool to teach haikuporter that when it is running the tests, it should examine `BUILD_PREREQUIRES` and `TEST_REQUIRES`. This would, of course, allow us to remove the hack above.
(This hack does not appear like this in any recipe, although app-arch/tar uses a `if [ "$recipeAction" = TEST ]; then` to append lines to BUILD_PREREQUIRES.)

I read _listBuildDependencies in Main.py, and:
* resolveBuildDependencies
* _doBuildStage
* _doTestStage

in Port.py but did not figure out how to modify haikuporter to teach it how to handle a TEST_REQUIRES.

BTW, I don't know what name would be most appropriate, between `TEST_REQUIRES` and `TEST_PREREQUIRES`.

Since haikuporter has received several great improvements recently by Python experts, I decided to file this issue.
Thanks!

**Comment by alaviss** [comment]
We really need this. A port I'm making (atf) requires (kyua) for it's `TEST()`, but the latter also requires the former to build

**Comment by fbrosson** [comment]
When I wrote the description of this issue I though that we would need to choose only one variable name, and I was suggesting to choose between `TEST_REQUIRES` and `TEST_PREREQUIRES`.
Well, it turns out that we need both of them.

I also thought we could workaround the chicken/egg issue by using a
```
if [ "$recipeAction" = TEST ]; then
	BUILD_REQUIRES="$BUILD_REQUIRES
		$TEST_REQUIRES
		"
	BUILD_PREREQUIRES="$BUILD_PREREQUIRES
		$TEST_PREREQUIRES
		"
fi
```
because I had noticed that it used to work in 2016 and/or 2017.
Unfortunately, for some reason unknown to me, this has stopped working in 2017 or maybe in 2016.

BTW, the `if [ "$recipeAction" = TEST ]; then ... fi` test still works. What seems to have changed is that the package activation only happens at build time, while it used to also take place at "test" time.

For now I suggest that, in recipes that have extra dependencies for running the tests, we just add the following comment and one or both of the next lines:
```
# Uncomment the following line before the build if you wish to run the tests.
#BUILD_REQUIRES+="$TEST_REQUIRES"
#BUILD_PREREQUIRES+="$TEST_PREREQUIRES"
```

**Comment by mmlr** [comment]
PREREQUIRES vs. REQUIRES are related to cross compilation, where PREREQUIRES would generally be run on the host (a cross compiler for example) and REQUIRES may be built for the target. Tests can generally not be run for cross built software, so using TEST_PREREQUIRES does not make much sense IMO.

Recipes are evaluated once and then cached in the recipe cache. Making the evaluated outcome of the recipe keys depend on the phase will therefore not work predictably. Such uses should be forbidden and removed if they are currently present anywhere.

**Comment by fbrosson** [comment]
I think we need both `TEST_PREREQUIRES` and `TEST_REQUIRES`.
This would be useful for some recipes of haikuports/haikuports/pull/2026 :

* `dev-lua/lutok/lutok-0.4.recipe` which has:
```
TEST_REQUIRES="
       devel:libatf_c++$secondaryArchSuffix
       "
TEST_PREREQUIRES="
       cmd:atf_sh$secondaryArchSuffix
       cmd:kyua$commandSuffix
       "
BUILD_REQUIRES+="$TEST_REQUIRES"
BUILD_PREREQUIRES+="$TEST_PREREQUIRES"
```
* ` dev-util/kyua/kyua-0.13.recipe` which has:
```
TEST_REQUIRES="
	devel:libatf_c++$secondaryArchSuffix
	"
TEST_PREREQUIRES="
	cmd:find
	cmd:sqlite3
	"
BUILD_REQUIRES+="$TEST_REQUIRES"
BUILD_PREREQUIRES+="$TEST_PREREQUIRES"
```
* `dev-util/pkgconf/pkgconf-1.3.90.recipe` which has:
```
TEST_REQUIRES="
+	devel:libatf_c$secondaryArchSuffix
+	"
+TEST_PREREQUIRES="
+	cmd:kyua$commandSuffix
+	"
+BUILD_REQUIRES+="$TEST_REQUIRES"
+BUILD_PREREQUIRES+="$TEST_PREREQUIRES"
```

I do agree that we should normaly build everything in `BUILD()`. The reason why some compilations can only happen in `TEST()` is because these _extra_ compilations depend on other recipes which depend on the compilations we are doing in `BUILD()`.
In other words, the chicken/egg dependencies can only be solved if we have both `TEST_PREREQUIRES` and `TEST_REQUIRES`.

The `if [ "$recipeAction" = TEST ]; then ... fi` test works fine and did allow, some time ago, to have extra dependencies activated at "test" time.
BTW, if you put `echo "recipeAction=$recipeAction"` in `BUILD()` and in `TEST()` you will get different results for `haikuporter` and  `haikuporter --test`.

Now, I have to admit that changing haikuporter is not easy at all. I've been trying to enhance it but did not succeed...

**Comment by mmlr** [comment]
PREREQUIRES are used for commands because they are used to build the recipe, so in a cross compile will be host tools running on the arch of the host. REQUIRES on the other hand are dependencies of the code, i.e. other cross compiled packages providing target arch libraries to link against. Tests are almost always run with the compiled binaries, and they can't run on the host, they need to run on the target. Consequently anything in TEST are REQUIRES and not PREREQUIRES. PREREQUIRES and REQUIRES aren't used to determine build order, so interdependencies aren't declared through them (the two levels wouldn't be enough anyway).

Regarding changing the parsed recipe output depending on recipeAction: With the current architecture it would mean to remove the recipe cache, which would have a huge performance impact. Recipe parsing is the most time consuming process in HaikuPorter, so aggresive caching is really desirable. To make this work again (i.e. the same as before the cache was introduced) there would need to be recipeAction specific caches. That still increases the overhead threefold for a rather questionable benefit, so I am against that.

Changing HaikuPorter isn't all that hard, but it needs to be done with a bit of care to not break other usecases.

**Comment by fbrosson** [comment]
OK; thanks.
So, unless haikuporter is changed to support recipeAction-specific caches (which I agree might not be a good idea), we will need to either:
1) comment out the tests which hit the chicken/egg issue, or
2) define `TEST_REQUIRES` and/or `TEST_PREREQUIRES` but comment out these lines:
```
#BUILD_REQUIRES+="$TEST_REQUIRES"
#BUILD_PREREQUIRES+="$TEST_PREREQUIRES"
```

(I think scenario 2 is better than scenario 1.)

I see your point regarding the `TEST_REQUIRES` vs `TEST_PREREQUIRES` topic, but I guess it would be OK to define both of these variables, since it allows to easily append the correct entries to `BUILD_REQUIRES` and `BUILD_PREREQUIRES`.

**Comment by fbrosson** [comment]
This is great news! Thanks @mmlr! I'll pull your test_requires branch and play with it asap.

BTW, we have a few recipes which define (but don't use) `TEST_PREREQUIRES`:
* curl-7.59.0
* gnutls36-3.6.2

We also have a few other recipes which define `TEST_REQUIRES` or `TEST_PREREQUIRES` (but also add it to `BUILD_REQUIRES` or `BUILD_PREREQUIRES`):
* unshield 1.4.2
* mandoc-1.14.3
* hefur-0.6
* primegen-0.97

Eureka! m4-1.4.18 defines `TEST_REQUIRES` and does not use it. So this can be tested "out of the box" :)
(I'll probably start the tests with atf, kyua and m4.)

It looks like these are currently the only recipes with a `TEST_[PRE]REQUIRES` already in master.
Thanks for your devs. I'll come back as soon as I have finished a few tests.

**Comment by mmlr** [comment]
As explained above, TEST_PREREQUIRES does not make much sense, hence I only added TEST_REQUIRES. Since the workaround of adding them to BUILD_* is not needed anymore, they can now simply be folded into the single TEST_REQUIRES.

**Comment by fbrosson** [comment]
I've tested your `test_requires` branch of haikuporter and it seems your devs are almost OK (or at least going to the right direction.)
I am able to use haikuporter's `--test` on recipes that have a `TEST_REQUIRES`, and haikuporter correctly activates everything mentioned in `TEST_REQUIRES`.

The small problem we have right now is that it does not _also_ activate what is in `BUILD_PREREQUIRES` and `BUILD_REQUIRES`.

For kyua, for example, once I changed `TEST_REQUIRES` to:
```
TEST_REQUIRES="
	$BUILD_REQUIRES
	$BUILD_PREREQUIRES
	cmd:find
	cmd:sqlite3
	"
```
`haikuporter --test kyua` worked like a charm :)

But this is just a workaround. haikuporter should automatically include `$BUILD_REQUIRES` and `$BUILD_PREREQUIRES` for us. This is important for recipes that do not have any `TEST_REQUIRES`.

BTW, there is a great news: `--enter-chroot`, combined with `--test`, works (almost) as expected. I'm saying almost because once the issue discussed above is fixed, `haikuporter --enter-chroot --test <package>` will work just as it will be supposed to :)

Regarding the `TEST_PREREQUIRES`, I will fix all the recipes that have this in order to only use `TEST_REQUIRES`.

Thanks for your great dev!

**Comment by mmlr** [comment]
That ```BUILD_*``` are not pulled in is by design and I am not convinced that they should. The reasoning behind not activating build requirements would be that some build systems will build the tests during their normal build phase and the test phase really only runs them. If that is not the case for some recipes, then adding ```$BUILD_*``` to ```TEST_REQUIRES``` would be appropriate and obvious.

This assumption is based on my past experience of build systems, but I don't really have a broad overview. So if it is more common to only build tests during the test phase, then implicitly adding the ```BUILD_*``` would probably make things easier.

Leaving out the build requires during the test phase would on the other hand ensure that the built binaries do really only depend on their runtime requirements. Then again, if this was a common problem I guess we would have seen that by now, as they would eventually fail when normally installed anyway.

**Comment by mmlr** [comment]
I've pushed the required changes to include the ```BUILD_*``` in the test phase (and some cleanup) to that branch for you to try out.

**Comment by fbrosson** [comment]
Excellent! With your latest changes I am able to test any package.
It works with recipes which have a `TEST_REQUIRES` (e.g. `kyua`, `atf`, `m4`) as well as with recipes which do not have a `TEST_REQUIRES` (e.g. `libmspack`, `mbedtls`)
(I already tested the examples above, and even a few others too.)

So, I guess I can update the following recipes:
* curl-7.59.0
* gnutls36-3.6.2
* unshield 1.4.2
* mandoc-1.14.3
* hefur-0.6
* primegen-0.97

with these changes:
```diff
-TEST_PREREQUIRES="
+TEST_REQUIRES="
```
```diff
-BUILD_REQUIRES+="$TEST_REQUIRES"
```
```diff
-BUILD_PREREQUIRES+="$TEST_PREREQUIRES"
```

This way it will be easier for everyone else to try your `test_requires` branch and provide you any additional feedback in case I missed something. (Well, I'm pretty sure everything is OK.)

**Comment by mmuman** [comment]
Couldn't pulling everything in create some circular dependencies sometimes? Just thinking…

**Comment by pulkomandy** [comment]
This is exactly why the TEST_REQUIRES have been split out from the BUILD_REQUIRES, yes (you may test software using something that depends on it). But in the other direction I think it should be fine.

**Comment by fbrosson** [comment]
@mmlr, your dev is just what we need; because now we can:
* build atf, lutok, lyua and pkgconf, (following this order will work)
* and then test these packages in any order, as long as they are installed (or available in `/path/to/haikuports/packages/`, of course.)

So thanks for you nice development!

**Comment by mmlr** [comment]
> build atf, lutok, lyua and pkgconf (using this order will work fine)

Note that using ```--{all|update}-dependencies``` the order in which these are built should not matter. Unmet dependencies should then simply be pulled in and built.

I have also pushed some more changes that will cause ports to be built automatically when testing was requested. This just seemed more helpful than the message telling the user to build the port first.

So ```haikuporter --update-dependencies --test atf``` should for example work out of the box even from a fresh portstree.

**Comment by fbrosson** [comment]
That's cool :)

Well, in my case I need to build each package one by one, because I'm using a ramdisk for my `haikports.conf`'s `OUTPUT_DIRECTORY`, so haikuporter refuses to move the hpkgs from `work-*/hpkgs/` to my `PACKAGES_PATH` (which is not in the ramdisk, of course.)
But, after all, this stop is nice because it allows me to go to the hpkgs directory and apply some custom scripts I wrote:
a) `drophrev`, which gets rid of the versioned dependency on the haiku package. This can be seen as a bad or risky practice, but it can be handy if I later choose to rollback to a previous Haiku hrev.
b) `dropliba` which deletes all the static libs in the packages passed as parameters.
(Obviously, these scripts use `package extract`, and `package create`, so running them in my ramdisk is convenient.)
Then I generally install the package and move it to my `PACKAGES_PATH`.

BTW, maybe I should investigate to see if hook scripts could be used with haikuporter. I haven't seen such a feature yet, but I haven't searched either :)

**Comment by pulkomandy** [comment]
The handling of hrevs will probably change a little once we have a stable release to refer to. That is, the packages may then depend on ">= beta1" rather than a specific hrev once things are stabilized.

As for static libraries, I think in most cases we don't need to keep them in packages. I don't care too much because they should be in _devel packages so they don't get in the way for users, at least.

**Comment by mmlr** [comment]
> a) drophrev, which gets rid of the versioned dependency on the haiku package. This can be seen as a bad or risky practice, but it can be handy if I later choose to rollback to a previous Haiku hrev.

This is indeed risking that the built package is not compatible with the downgraded hrev. A cleaner way to build against an older hrev is to put the corresponding ```haiku_*``` HPKGs into the ```packages``` directory. This way the packages are actually built against that hrev. This works as long as the host Haiku is syscall compatible with the one used inside the chroot (which most often is the case as the syscall and kernel interfaces don't change often).

**Comment by fbrosson** [comment]
@mmlr, there's a small issue.

**Comment by fbrosson** [comment]
I had noticed that you had added the package itself as a test dependency, but it looks like that's not OK in some cases. So I think it would be better to not add the patckage itself.

I found this just 2 minutes ago. I've updated my x86_gcc2 system to the latest hrev, and rebuilt git 2.16.3 to see if the tests were OK. But I got this error:
```
TESTING dev-vcs::git-2.16.3
----------------------------------------------------------------------
Cleaning up temporary directories ...
Fetching package for alien_svn ...
*** Failed to find a match for "alien_svn": Name not found
```
That's because alien_svn is not available. @pulkomandy has recently [added](https://github.com/haikuports/haikuports/commit/0a1ac6b520c2c53138addab4cc2fd264831a4fa9) it as a dependency of git_svn. But [alien_svn](https://github.com/haikuports/haikuports/blob/084cbd887cb288e241900c924152c0a06421cf84/dev-perl/alien_svn/alien_svn-1.8.11.0.recipe) is only available on x86_gcc2.
BTW, I think I have tried to build alien_svn, but I guess it failed to build.

Anyway, it looks like we need an extra small change in haikporter's new feature.

Hmm, I'm going to test the change you just pushed to your repo one hour ago.

**Comment by mmlr** [comment]
The base package is not added by HaikuPorter during the test phase explicitly. In most ports it will however be pulled in due to the base package requires of the subpackages. This isn't really avoidable because these are explicit requires and they are part of the dependency infos, which in turn are the basis for dependency resolution. Since they are part of the normal requires declaration, these can't be filtered out easily either.

The failure to test git does not really come from that dependency however. The git build is simply broken due to the missing alien_svn. I have not looked at x86_gcc2, but for x86_64 alien_svn simply does not exist. So this isn't really related to the TEST_REQUIRES changes.

**Comment by fbrosson** [comment]
Well, if I use the previous haikuporter, I can both build and test git (even without the alien_svn.)

With the new haikuporter, I can only build it. (If I try to test it then I get the error.)

So I'd say that, in `--test` mode, haikuporter should only pull what is in `BUILD_REQUIRES`, `BUILD_PREREQUIRES` and `TEST_REQUIRES`.
It should not pull what is in `REQUIRES` or any `REQUIRES_*`.
If the tests need anything that is also in a `REQUIRES`, then it should be, imo, explicitly added to `TEST_REQUIRES`.
After all, when we are testing git, we're not testing the runtime from the package but the runtime from the build directory.

**Comment by fbrosson** [comment]
@mmlr, maybe we could create a _build package_ for the `TEST` phase, the same way it is done for the `BUILD` phase. We would get:
```
waiting for build package git-2.17.0-1 to be activated
```
and that fake package would depend on no other package.
Do you think this would be possible? Thanks!

**Comment by fbrosson** [comment]
Hmm, it looks like the build package is already activated in the TEST phase.
(I'll delete the previous message and this one. Sorry for the noise.)

**Comment by mmlr** [comment]
That's an interesting case actually and it brings up a couple of issues.

IMO git shouldn't even build with one of its runtime dependencies missing. First of all, the affected subpackage won't be installable, but you don't necessarily notice that up front. And second, the RequiresUpdater is not able to do its job in that case and therefore won't ensure that a tested minimal version requirement is recorded in the package.

Only the "no-dependencies" modes, i.e. the default and ```--get-dependencies``` modes will actually allow to build in such a case. The ```--all-dependencies```, ```--update-dependencies``` and the new ```--missing-dependencies``` modes (as well as the special buildmaster mode) always do a full dependency resolution up front, including the declared runtime dependencies. In these modes, the git build already failed for that reason. I always used ```--update-dependencies``` and now ```--missing-dependencies``` and therefore didn't even notice the discrepancy.

You are correct that it is a change in behaviour, previously ```--test``` indeed would not pull in runtime dependencies. The reasoning against doing so "we are not testing the runtime from the package but from the build directory" doesn't really hold though. It is reasonable to expect the tests of a project to have similar runtime dependencies as the actual runtime. If there was a test for ```git-svn``` for example, it would most certainly require ```alien_svn```.

I'm a bit undecided on this one. Removing the ```REQUIRES``` from the ```--test``` dependency resolution is quite literally just that. It simple, it will work and it will restore the previous behaviour. But IMO the previous behaviour is simply broken and that brokenness should be addressed.

**Comment by fbrosson** [comment]
Well, I believe the best way to address this issue is to remove the `REQUIRES` from the `--test` and require that any dependency required for the tests be explicitly added to `TEST_REQUIRES`.
(This would be great ;)

BTW, it would be cool to have an option to tell haikuporter that if something in `TEST_REQUIRES` is missing then it should still kick the test. This option could be called `--ignore-missing-dependencies` or anything else. Such an option would be handy if a particular dependency turns out to be broken.
(Shame on me! You have just added the great `TEST_REQUIRES` feature and now I am already suggesting new features!!!)
